### DISCUSSION TITLE
Ways to implement continuous looping of wav file with no gaps or delays

### DISCUSSION THREAD

**Discussion by timmcarson** [pschatzmann/arduino-audio-tools#768] [Q&A] [answered]
As the tittle suggests, I am looking for ways to implement continuous looping of a single wav file with no gaps or delays. 

I am using AudioPlayer.h 

So if the ending block has data for 300 out of 1024 I want to fill the remainder with the beginning of the file.  I would appreciate any suggestions.

Timm

**Comment by pschatzmann** [comment, marked as answer]
a) You would implement a subclass of AudioStream that implements your looping logic using a file as input.
b) You would implement an AudioSource (e.g. using the AudioSourceCallback) that returns a)

**Comment by timmcarson** [comment]
Thanks,

I will work on that.

**Reply by pschatzmann** [reply]
a) should be pretty easy to implement for WAV files: at the end you would just do a file.seek(44) to position to the beginning of the pcm audio data.

**Comment by pschatzmann** [comment]
With the latest [commit](https://github.com/pschatzmann/arduino-audio-tools/commit/fa7c5c03f32fef8330b21eef512f774169b8ef62) I have added a file looper with an example. I suggest that you also read the [class documentation](https://pschatzmann.github.io/arduino-audio-tools/classaudio__tools_1_1_file_loop.html) and have a look at the example.

For the AudioSourceCallback there is already an example that you can use as starting point.

**Reply by timmcarson** [reply]
Hi @pschatzmann ,

I was able to successfully run the file looper example.  

I tried to adapt the AudioSourceCallback example (player-callback-i2s.ino)  to use audiokitstream but does not play.  Can you tell what I am doing wrong?

my code:

<details>
  <summary>My Code</summary>



```cpp

#include <Arduino.h>
/**
 * @file player-sd-callback.ino
 * @brief see https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-player/player-callback-i2s/README.md
 * 
 * @author Phil Schatzmann
 * @copyright GPLv3
 */


// Install https://github.com/pschatzmann/arduino-libhelix.git

#include "Arduino.h"
#include "AudioTools.h"
#include "AudioLibs/AudioKit.h"// Added
#include "AudioLibs/FileLoop.h"// Added
#include "AudioCodecs/CodecWAV.h"// Added
#include <SPI.h>
#include <SD.h>


// forward declarations
void callbackInit();
Stream* callbackStream(int offset);


// data
const int chipSelect=PIN_AUDIO_KIT_SD_CARD_CS;  // Changed to audio kit pin
AudioSourceCallback source(callbackStream,callbackInit);
AudioKitStream i2s;  // Changed to audio kit
//I2SStream i2s;
WAVDecoder decoder;  // Changed to Wav decoder
AudioPlayer player(source, i2s, decoder);
File audioFile;



void callbackInit() {
  SD.begin(chipSelect);
  audioFile = SD.open("/");
}

Stream* callbackStream(int offset) {
  auto lastFile = audioFile;
  audioFile = audioFile.openNextFile();
  lastFile.close();
  return &audioFile;
}

void callbackPrintMetaData(MetaDataType type, const char* str, int len){
  Serial.print("==> ");
  Serial.print(toStr(type));
  Serial.print(": ");
  Serial.println(str);
}


void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);

  // setup output
  auto cfg = i2s.defaultConfig(TX_MODE);
  cfg.sd_active = true;
  i2s.begin(cfg);

  // setup player
  player.setMetadataCallback(callbackPrintMetaData);
  player.begin();
}

void loop() {
  player.copy();
}
```

</details>

<details>
  <summary>Serial Monitor Output</summary>

```

ets Jun  8 2016 00:22:57

rst:0x1 (POWERON_RESET),boot:0x37 (SPI_FAST_FLASH_BOOT)
flash read err, 1000
ets_main.c 371 
ets Jun  8 2016 00:22:57

rst:0x10 (RTCWDT_RTC_RESET),boot:0x37 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:1184
load:0x40078000,len:13192
load:0x40080400,len:3028
entry 0x400805e4
E (219) esp_core_dump_flash: Core du�␆␂"�х�check failed:
Calculated checksum='de8eb97c'
Image checksum='ffffffff'
[    19][D][esp32-hal-cpu.c:244] setCpuFrequencyMhz(): PLL: 480 / 2 = 240 Mhz, APB: 80000000 Hz
[I] AudioTypes.h : 78 - sample_rate: 44100
[I] AudioTypes.h : 79 - channels: 2
[I] AudioTypes.h : 80 - bits_per_sample: 16
[I] I2SConfig.h : 106 - rx/tx mode: TX_MODE
[I] I2SConfig.h : 107 - port_no: 0
[I] I2SConfig.h : 108 - is_master: Master
[I] I2SConfig.h : 109 - sample rate: 44100
[I] I2SConfig.h : 110 - bits per sample: 16
[I] I2SConfig.h : 111 - number of channels: 2
[I] I2SConfig.h : 112 - i2s_format: I2S_STD_FORMAT
[I] I2SConfig.h : 114 - auto_clear: true
[I] I2SConfig.h : 116 - use_apll: true
[I] I2SConfig.h : 121 - buffer_count:6
[I] I2SConfig.h : 122 - buffer_size:512
[I] I2SConfig.h : 126 - pin_bck: 14
[I] I2SConfig.h : 128 - pin_ws: 15
[I] I2SConfig.h : 130 - pin_data: -1515870811
[W] AudioKit.h : 58 - sd_active = true
[   102][I][esp32-hal-i2c.c:75] i2cInit(): Initialising I2C Master: sda=18 scl=23 freq=100000
[I] I2SConfig.h : 106 - rx/tx mode: TX_MODE
[I] I2SConfig.h : 107 - port_no: 0
[I] I2SConfig.h : 108 - is_master: Master
[I] I2SConfig.h : 109 - sample rate: 44100
[I] I2SConfig.h : 110 - bits per sample: 16
[I] I2SConfig.h : 111 - number of channels: 2
[I] I2SConfig.h : 112 - i2s_format: I2S_STD_FORMAT
[I] I2SConfig.h : 114 - auto_clear: true
[I] I2SConfig.h : 116 - use_apll: true
[I] I2SConfig.h : 121 - buffer_count:6
[I] I2SConfig.h : 122 - buffer_size:512
[I] I2SConfig.h : 126 - pin_bck: 5
[I] I2SConfig.h : 128 - pin_ws: 25
[I] I2SConfig.h : 130 - pin_data: 26
[I] AudioKit.h : 591 - void audio_tools::AudioKitStream::setupActions()
[I] AudioKit.h : 341 - void audio_tools::AudioKitStream::addAction(int, void (*)(bool, int, void*), void*)
[I] AudioActions.h : 42 - ActionLogic::add pin: 39 / logic: 0
[I] AudioActions.h : 52 - pin 39 -> INPUT_PULLUP
[I] AudioActions.h : 42 - ActionLogic::add pin: 19 / logic: 2
[I] AudioActions.h : 55 - pin 19 -> INPUT
[W] AudioKit.h : 616 - Volume Buttons ignored because of conflict: 13
[I] AudioPlayer.h : 408 - virtual void audio_tools::AudioPlayer::setMetadataCallback(void (*)(audio_tools::MetaDataType, const char*, int), audio_tools::ID3TypeSelection)
[I] AudioPlayer.h : 351 - setVolume(1.000000)
[I] VolumeStream.h : 191 - setVolume: 1.000000
[I] VolumeStream.h : 191 - setVolume: 1.000000
[I] AudioTypes.h : 78 - sample_rate: 44100
[I] AudioTypes.h : 79 - channels: 2
[I] AudioTypes.h : 80 - bits_per_sample: 16
[I] MetaDataID3.h : 564 - virtual void audio_tools::MetaDataID3::begin()
[W] AudioPlayer.h : 185 - -> begin: no data found
```

</details>

**Comment by pschatzmann** [comment]
Check the wiki of the https://github.com/pschatzmann/arduino-audiokit/wiki
And this: https://github.com/pschatzmann/arduino-audio-tools/wiki/It's-not-working

**Reply by timmcarson** [reply]
So I have verified that the board is working fine.  Both streams-file_loop-audiokit.ino and player-sd-i2s.ino execute as expected.  

I still can't get (player-callback-i2s.ino) to play the logger says "AudioPlayer.h : 185 - -> begin: no data found" but I can list the files on the SD card so I know the card is mounting properly.  is there anyway to get more details on why the AudioSourceCallback doesn't seem to be doing anything?

Thanks

**Reply by pschatzmann** [reply]
So, why don't you just log in the callbackStream what file you are returning? 
I guess it is not a valid audio file...

**Comment by timmcarson** [comment]
I think I found where the issue is.

the AudioPlayer 
starts the source and the callbackInit() runs 
then it sets:  p_input_stream = p_source->selectStream(index);

But in the AudioSourceCallback Class in AudioSource.h, it looks like "selectStream(index)" returns nullptr because  indexStreamCallback is nullptr

```    /// Returns selected audio stream
    virtual Stream* selectStream(int index) override {
        return indexStreamCallback == nullptr ? nullptr : indexStreamCallback(index);
    }```

I am not sure where / how the indexStreamCallbeack is set

**Reply by pschatzmann** [reply]
The indexStreamCallback is defined  by calling  setCallbackSelectStream() in the AudioSourceCallback class.
We could change the implementation so, that if this is not defined we just do a begin() and nextStream(n) to make this method optional

The initial AudioSource class was very lean, but it has been getting a bit tricky by merging the different pull requests from different users who wanted to have their use case reflected...

**Comment by pschatzmann** [comment]
Just committed the following logic which makes the selectStream optional:
```C++
    /// Returns selected audio stream
    virtual Stream* selectStream(int index) override {
        LOGI("selectStream: %d", index);
        if (indexStreamCallback==nullptr){
            LOGI("setCallbackSelectStream not provided");
            if (index>0) begin();
            return nextStream(index);
        }
        return indexStreamCallback(index);
    }
```
I also tested and updated the example

**Reply by pschatzmann** [reply]
Just one final comment: I don't quite get why somebody would want to use the Player for an endless looping file. This is somehow a contradiction because the player is just about moving automatically to the next file when the current file ends.
 
If you want to implement a file based looping you are better off using the FileLoop class!

**Reply by timmcarson** [reply]
Thanks for you help.

I tried the new example, using AudioKitStream and WAVDecoder since I have a lyrat board and wave files.

The example didn't compile for me, I think the forward declaration of the stream is missing "Next"

``` 
Stream* callbackStream(int offset);  
should be 
Stream* callbackNextStream(int offset);```

It then compiled, but stops at [I] StreamCopy.h : 74 - buffer_size=1024

```
[I] AudioKit.h : 596 - void audio_tools::AudioKitStream::setupActions()
[I] AudioKit.h : 346 - void audio_tools::AudioKitStream::addAction(int, void (*)(bool, int, void*), void*)
[I] AudioActions.h : 42 - ActionLogic::add pin: 39 / logic: 0
[I] AudioActions.h : 52 - pin 39 -> INPUT_PULLUP
[I] AudioActions.h : 42 - ActionLogic::add pin: 19 / logic: 2
[I] AudioActions.h : 55 - pin 19 -> INPUT
[W] AudioKit.h : 621 - Volume Buttons ignored because of conflict: 13
[I] AudioPlayer.h : 408 - virtual void audio_tools::AudioPlayer::setMetadataCallback(void (*)(audio_tools::MetaDataType, const char*, int), audio_tools::ID3TypeSelection)
[I] AudioPlayer.h : 351 - setVolume(1.000000)
[I] VolumeStream.h : 191 - setVolume: 1.000000
[I] VolumeStream.h : 191 - setVolume: 1.000000
[I] AudioTypes.h : 85 - sample_rate: 44100
[I] AudioTypes.h : 86 - channels: 2
[I] AudioTypes.h : 87 - bits_per_sample: 16
[I] MetaDataID3.h : 564 - virtual void audio_tools::MetaDataID3::begin()
[I] AudioSource.h : 96 - selectStream: 0
[I] AudioSource.h : 98 - setCallbackSelectStream not provided
[I] StreamCopy.h : 74 - buffer_size=1024
```

do you think there is an issue changing the decoder and stream?

The reason audio player and loop (might be wrong thinking).  is there are multiple file I want to be able to cycle through but only when I select a button to do so.

**Reply by pschatzmann** [reply]
I suggest that you open a new discussion and describe your exact requirements. "is there are multiple file I want to be able to cycle through but only when I select a button to do so". I don't think that the CallbackSource is the right solution for this. It seems you should be looking at [setAutoNext()](https://pschatzmann.github.io/arduino-audio-tools/classaudio__tools_1_1_audio_player.html#afdd194c661ce148c25ba122293e342d7)
